pkg/logs/input/docker/decoder.go

```diff
@@ -0,0 +1,74 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+// +build docker
+
+package docker
+
+import (
+	"bytes"
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/decoder"
+)
+
+// Decoder specific to docker container source.
+type Decoder struct {
+	*decoder.Decoder
+}
+
+// InitializeDecoder returns a properly initialized Decoder
+func InitializeDecoder(source *config.LogSource, containerID string) *Decoder {
+
+	return &Decoder{
+		Decoder: decoder.NewDecoderWithStrategy(source, NewParser(containerID), &matchHeader{}),
+	}
+}
+
+// Start starts the Decoder
+func (d *Decoder) Start() {
+	d.Decoder.Start()
+}
+
+// Stop stops the Decoder
+func (d *Decoder) Stop() {
+	d.Decoder.Stop()
+}
+
+const (
+	dockerHeaderLen       = 8
+	dockerHeaderPrefixLen = 4
+)
+
+var (
+	dockerHeaderStdoutPrefix = []byte{1, 0, 0, 0}
+	dockerHeaderStderrPrefix = []byte{2, 0, 0, 0}
+)
+
+type matchHeader struct {
+	decoder.EndLineStrategy
+}
+
+// EndLineCriterion does an extra checking on matching docker header. The header should be
+// ignored for determine weather it's a end of line or not.
+func (s *matchHeader) EndLineCriterion(buffer *bytes.Buffer, bs []byte, start int, end int) bool {
+	return bs[end] == '\n' && !s.matchDockerHeader(buffer, bs[start:end])
+}
+
+// When a newline (in byte is 10) is matching, an additional check need to
+// be done to make sure this is not part of docker header.
+// [1|2 0 0 0 size1 size2 size3 size4], where size1 size2 size3 size4 can be
+// 10 in byte.
+// case [1|2 0 0 0 10 size2 size3 size4]
+// case [1|2 0 0 0 sizeNot10 10 size3 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 10 size4]
+// case [1|2 0 0 0 sizeNot10 sizeNot10 sizeNot10 10]
+func (s *matchHeader) matchDockerHeader(buffer *bytes.Buffer, bs []byte) bool {
+	l := buffer.Len() + len(bs)
+	if l >= dockerHeaderLen || l < dockerHeaderPrefixLen {
+		return false
+	}
+	h := append(buffer.Bytes(), bs...)
```

Comment:
in most of the logs code base, we have full variable names, can we stick to it please ?